### PR TITLE
Test Suite: new test mode: Combined. Test a profile using test scenarios.

### PR DESCRIPTION
**Pull request by yuumasato**
#### Description:

- Enable testing of rules selected in a profile using ~~rule mode~~ test scenarios.
  ~~- Introduce a new argument `--profile` that will use last argument as a profile `ID`.~~
  - Adds scraper functions that return rules selected in a profile.
  - Adds a new checker `CombinedChecker`
    - The checker outputs message for every selected rule that was not tested
    - The checker overrides `# profiles` metadata in test scenarios and sets it as the target profile

#### Rationale:

- Allows batch testing with test scenarios for rules that are selected in a profile.
- Facilitate test scenario testing by overriding `# profiles` metadata, but with increased risk of test failures due to differences in variable selection and setup.

### CODE REVIEW

File: tests/test_suite.py [hunk 1]
```diff
@@ -121,7 +121,9 @@ def parse_args():
                                    "target is handled as a substring - so you can "
                                    "ask for subset of all rules this way. (If you "
                                    "type ipv6 as a target, all rules containing "
-                                   "ipv6 within id will be performed."))
+                                   "ipv6 within id will be performed."
+                                   "If --profile is specified, this argument "
+                                   "specifies the profile to be tested."))
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply tests/test_suite.py]: To have a mutually exclusive using `argparse`, the arguments must be all optional.
That would look like this:
```
usage: test_suite.py rule [-h]
                          (--docker BASE_IMAGE | --container BASE_IMAGE | --libvirt HYPERVISOR DOMAIN HYPERVISOR DOMAIN)
                          --datastream DATASTREAM
                          [--xccdf-id REF-ID | --xccdf-id-number REF-ID-SELECT]
                          [--loglevel LOGLEVEL] [--logdir LOGDIR]
                          [--mode {online,offline}]
                          [--remediate-using {oscap,bash,ansible}] [--debug]
                          [--dontclean] [--scenarios SCENARIOS_REGEX]
                          [--profile_target PROFILE_TARGET | --rule_target TARGET [TARGET ...]]
```
I'm not a fan of this approach as you'd have to specify `--rule-target` for the default behavior.
@jan-cerny Do you have a suggestion how to improve this?

[the remark this answers is not in the document]
Comment (yuumasato) [reply tests/test_suite.py]: I have tried to improve the argument help message, hopefully they are clearer.

File: tests/ssg_test_suite/common.py [hunk 2]
```diff
@@ -26,6 +26,9 @@
     "-o", "UserKnownHostsFile=/dev/null",
 ) + SSH_ADDITIONAL_OPTS
 
+CONTENT_BASE_PREFIX = "xccdf_org.ssgproject."
+CONTENT_PROFILE_PREFIX = CONTENT_BASE_PREFIX + "content_profile_"
+CONTENT_RULE_PREFIX = CONTENT_BASE_PREFIX + "content_rule_"
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply tests/ssg_test_suite/common.py]: Moved them to `constants.py`

File: tests/test_suite.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jan-cerny) [reply tests/test_suite.py]: @yuumasato I don't, I haven't realized that.

File: tests/test_suite.py [hunk 3]
```diff
@@ -136,6 +136,14 @@ def parse_args():
                              dest="scenarios_regex",
                              default=None,
                              help="Regular expression matching test scenarios to run")
+    parser_rule.add_argument("--source", choices=['rule', 'profile'],
+                              default='rule',
+                              dest="target_source",
+                              help=("Specify wether target is a list of rules, or a profile. "
```
Comment (jan-cerny) [inline tests/test_suite.py]: @yuumasato Please change it to "whether".

File: tests/ssg_test_suite/rule.py [hunk 4]
```diff
@@ -340,4 +341,25 @@ def perform_rule_check(options):
...
+        if not profile.startswith(OSCAP_PROFILE):
+            profile = OSCAP_PROFILE+profile
+        logging.info("Performing rule test using profile: {0}".format(profile))
+
+        # Fetch target list from rules selected in profile
+        target_rules = xml_operations.get_all_rule_ids_in_profile(
+                options.datastream, options.benchmark_id,
+                profile, logging)
+        logging.debug("Profile {0} expanded to following list of "
+                      "rules: {1}".format(profile, target_rules))
```
Comment (jan-cerny) [inline tests/ssg_test_suite/rule.py]: @yuumasato I have a question: What happens if some of the rules don't have any test scenarios?

**Comment by jan-cerny** [comment]
There is a confusing thing: You provide profile ID as an argument to ./test_suite.py, but it runs test scenarios for rules according to `# profile` directive in each scenario, and that can be a different profile. It gives a false impression that the rule is tested in a context of the profile.

```
./test_suite.py rule --libvirt qemu:///system ssgts_rhel8 \
--datastream ../build/ssg-rhel8-ds.xml \
--source profile cjis

....

INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script comment.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script line_not_there.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - xccdf_org.ssgproject.content_rule_sshd_disable_rhosts
```

**Comment by yuumasato** [comment]
> There is a confusing thing: You provide profile ID as an argument to ./test_suite.py, but it runs test scenarios for rules according to `# profile` directive in each scenario, and that can be a different profile. It gives a false impression that the rule is tested in a context of the profile.
> 
> ```
> ./test_suite.py rule --libvirt qemu:///system ssgts_rhel8 \
> --datastream ../build/ssg-rhel8-ds.xml \
> --source profile cjis
> 
> ....
> 
> INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
> INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
> INFO - Script comment.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
> INFO - Script line_not_there.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
> INFO - xccdf_org.ssgproject.content_rule_sshd_disable_rhosts
> ```

@jan-cerny But these are the normal outputs of "rule mode". Each test scenario targets a profile, thus, the message with  `using profile`.

File: tests/ssg_test_suite/rule.py (hunk 4 above)
Comment (yuumasato) [reply tests/ssg_test_suite/rule.py]: The rule won't be tested.

**Comment by yuumasato** [comment]
> It gives a false impression that the rule is tested in a context of the profile.

would it make sense then for the Test Suite to override the `#profile` metadata, when running "rule mode" on rules selected on a profile? 
As the rule is already selected in the argument profile, this shouldn't be a problem.
But there would be an issue with any variable specifics that the test scenario sets up.

File: tests/ssg_test_suite/rule.py (hunk 4 above)
Comment (yuumasato) [reply tests/ssg_test_suite/rule.py]: I will look into writing a message when a rule is not tested.

File: tests/test_suite.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matejak) [reply tests/test_suite.py]: What about making this a new subcommand? The rule mode does what it says - you just specify the rule regex, and it tests it. This improvement adds complexity to the command-line invocation, and even more complexity is coming to the rule mode itself (We are for example thinking of narrowing down scenarios, so if you are fighting with a failing one, you get faster test suite feedback on your changes by skipping non-relevant scenarios.)
I believe that a new subcommand can be introduced with quite low overhead.

**Comment by matejak** [comment]
> would it make sense then for the Test Suite to override the `#profile` metadata, when running "rule mode" on rules selected on a profile?

This makes sense, and it is one more reason why this great improvement should be separated from the rule subcommand.
Regarding variables, some may be worked around (e.g. timeouts - you write tests for 1 seconds vs 1 year, and they will pass/fail in the same fashion in all profiles), and hopefully those problematic cases won't be many, so things may be kept under control.

On a slightly related note, we should be able to use the `(all)` profile now as soon as it gets fixed in the test suite / oscap-ssh.

**Comment by jan-cerny** [comment]
> why this great improvement should be separated from the rule subcommand.

Do you mean introducing a third mode next to to rule and profile? I like that, it is friendlier for the users than "profile target of rule mode".

**Comment by jan-cerny** [comment]
> would it make sense then for the Test Suite to override the #profile metadata, when running "rule mode" on rules selected on a profile? 

This override makes sense also when you test a single rule, but you want to do it in a context of different profile than specified in the test scenario.

**Comment by yuumasato** [comment]
> Do you mean introducing a third mode next to to rule and profile? I like that, it is friendlier for the users than "profile target of rule mode".

What if I call this mode "profile target of rule mode"?
Jokes apart, suggestion on how to call this mode?

**Comment by jan-cerny** [comment]
We will have 3 modes:
1. runs test scenarios only for selected rules - "single-rule"
2. runs test scenarios for all rules in the given profile - "all-rules"
3. runs whole profile, doesn't run test scenarios - "one-pass"

File: tests/test_suite.py (hunk 3 above)
Comment (yuumasato) [reply tests/test_suite.py]: This doesn't apply anymore.

File: tests/ssg_test_suite/rule.py (hunk 4 above)
Comment (yuumasato) [reply tests/ssg_test_suite/rule.py]: The `CombinedChecker` will output a list of rules not tested at the end of its run.

**Comment by yuumasato** [comment]
@jan-cerny @matejak Thank you for your feedback.
I have reworked the whole thing, and updated the PR description.

**Comment by jan-cerny** [comment]
@yuumasato Thanks, I'll take a look.

**Comment by jan-cerny** [comment]
Works great for me!

**Pull request merged**